source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -50,11 +53,111 @@ void Filter::onRead() {
 
 void Filter::onReadWorker() {
   Network::ConnectionSocket& socket = cb_->socket();
-  std::string proxy_line;
-  if (!readLine(socket.fd(), proxy_line)) {
+  WireHeader hdr;
+
+  if (!readProxyHeader(socket.fd(), hdr)) {
     return;
   }
 
+  if (hdr.valid_.has_value()) {
+    // Error check the source and destination fields. Most errors are caught by the address
+    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
+    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
+    // address.
+    const auto remote_version = hdr.remote_address->ip()->version();
+    const auto local_version = hdr.local_address->ip()->version();
+    if (remote_version != hdr.protocol_version || local_version != hdr.protocol_version) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+    // Check that both addresses are valid unicast addresses, as required for TCP
+    if (!hdr.remote_address->ip()->isUnicastAddress() ||
+        !hdr.local_address->ip()->isUnicastAddress()) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+
+    // Only set the local address if it really changed, and mark it as address being restored.
+    if (*hdr.local_address != *socket.localAddress()) {
+      socket.setLocalAddress(hdr.local_address, true);
+    }
+    socket.setRemoteAddress(hdr.remote_address);
+  }
+
+  // Release the file event so that we do not interfere with the connection read events.
+  file_event_.reset();
+  cb_->continueFilterChain(true);
+}
+
+void Filter::parseV2Header(char* buf, size_t len, WireHeader& hdr) {
+  // Skip the first 12-bytes
+  // Next is ver/cmd
+  const int ver_cmd = buf[PROXY_PROTO_V2_SIGNATURE_LEN];
+  // Only do connections on behalf of another user, not
+  // internally-driven health-checks. If its not on behalf
+  // of someone, or its not AF_INET{6} / STREAM/DGRAM, ignore
+  // and use the real-remote info
+  if ((ver_cmd & 0xf) == PROXY_PROTO_V2_ONBEHALF_OF) {
+    uint8_t proto_family = buf[PROXY_PROTO_V2_SIGNATURE_LEN + 1];
+    if ((((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET ||
+         ((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET6) &&
+        (((proto_family & 0x0f) == PROXY_PROTO_V2_TRANSPORT_STREAM) ||
+         ((proto_family & 0x0f) == PROXY_PROTO_V2_TRANSPORT_DGRAM))) {
+      hdr.valid_ = true;
+      if (((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET) {
+        typedef struct {
+          uint32_t src_addr;
+          uint32_t dst_addr;
+          uint16_t src_port;
+          uint16_t dst_port;
+        } __attribute__((packed)) pp_ipv4_addr;
+        pp_ipv4_addr* v4;
+        if (len < PROXY_PROTO_V2_HEADER_LEN + sizeof(pp_ipv4_addr))
```

Comment:
braces around if statements, please check the entire change in case i'm missing anything.